refactor(api-docs): rework ApiExplorerPage to use EntityListProvider

kuangp opened the pull request:
Signed-off-by: Phil Kuang <pkuang@factset.com>

## Hey, I just made a Pull Request!
Rework `ApiExplorerPage` to ultilize `EntityListProvider` to provide a consistent UI with the `CatalogIndexPage` which now exposes support for starring entities, pagination, and customizing columns. 

Closes #4637

![Screen Shot 2021-06-10 at 12 59 19 PM](https://user-images.githubusercontent.com/6998196/121568012-35ae9c80-c9ed-11eb-88ba-e60dbb4e61b5.png)

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

timbonicus commented on plugins/api-docs/src/components/ApiExplorerTable/ApiExplorerTable.tsx:
```diff
@@ -1,214 +0,0 @@
-
-const filters: TableFilter[] = [
-  {
-    column: 'Owner',
```
This is losing functionality in that the `owner` and `lifecycle` filters are removed.

kuangp replied:
Ah, good point, I'll look into implementing these

timbonicus reviewed (commented):
Hey, this is a really awesome contribution! 😍 So cool to see a continuation of the `EntityListProvider` work.

The big concern here, I think, is losing the `lifecycle` and `owner` filters that exist today on the api-docs page. Of course these filters _also_ make sense on the CatalogPage, so I'd expect to create them in `catalog-react` alongside the other Picker components, so they could be used in both places. These could possibly register simple frontend filters for now, especially as **backend** `owner` filtering depends on #4499.

And - taking that mentally one step further - there's some interesting overlap with the _existing_ `OwnerPicker` component in the `scaffolder` plugin. Maybe not enough to sensibly consolidate these, but they should at least borrow from each other I would imagine.

To mentally go _even further_,  the ApiDocsPage _itself_ should become unnecessary once there's a UI for `EntityKindPicker` available on the CatalogPage. This makes sense as an interim step, though.

timbonicus commented:
Noting too that I'm not 100% sure this closes #5310, since that mentions using the pagination exposed by `catalog-backed`. The catalog page, and therefore this PR as well, do client-side pagination rather than backend pagination.

kuangp commented:
> Noting too that I'm not 100% sure this closes #5310, since that mentions using the pagination exposed by `catalog-backed`. The catalog page, and therefore this PR as well, do client-side pagination rather than backend pagination.

good point, I'll remove that issue from this PR

kuangp commented:
@timbonicus updated this change with implementations for lifecycle and owner filters

timbonicus reviewed (approved):
Thanks Phil! This is looking really good, thanks for adding those missing filters. I hope you don't mind, I added one commit - primarily to add some spacing around the filters, and some minor language tweaks. I also added these new filters to the `CatalogPage` 😍 

One more minor comment -- and we'll need @backstage/maintainers for final approval. I think this is just about ready to go, though!

kuangp commented:
> Thanks Phil! This is looking really good, thanks for adding those missing filters. I hope you don't mind, I added one commit - primarily to add some spacing around the filters, and some minor language tweaks. I also added these new filters to the `CatalogPage` 😍
> 
> One more minor comment -- and we'll need @backstage/maintainers for final approval. I think this is just about ready to go, though!

thanks for the updates!

kuangp commented on plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx:
```diff
@@ -0,0 +1,83 @@
+      [
+        ...new Set(
+          backendEntities
+            .map((e: Entity) => e.spec?.owner)
```
that definitely makes sense, let me know if we want to accommodate this now or if we're ok with shipping the implementation as is and enhancing later.

timbonicus replied:
Brought this up with maintainers, would rather switch to using relations for this before merge 🙏

kuangp replied:
@timbonicus updated the `EntityOwnerPicker` to utilize the `ownedBy` relation

kuangp commented:
@timbonicus really sorry but I rebased my branch and forced pushed without pulling down your commit and accidentally erased your changes. Do you still happen to have the commit locally to re-push onto this branch?

Rugvip commented on plugins/api-docs/src/components/ApiExplorerPage/ApiExplorerPage.tsx:
```diff
@@ -18,58 +18,68 @@ import {
-              Register Existing API
-            </Button>
-          )}
+          <CreateComponentButton label="Register Existing API" />
```
Tbh I actually think it's best to keep the existing implementation in place here instead of re-using the button from the catalog plugin. It's not a log of code, and we end up with an individual switch and external route ref for the api docs plugin, which with this gets coupled together with the catalog plugin.

Rugvip commented on plugins/catalog/src/index.ts:
```diff
@@ -18,6 +18,7 @@ export * from './components/AboutCard';
 export { CatalogLayout } from './components/CatalogPage';
 export { CatalogResultListItem } from './components/CatalogResultListItem';
 export { CatalogTable } from './components/CatalogTable';
+export type { EntityRow } from './components/CatalogTable';
```
I'd vote for renaming this to `CatalogTableRow` as part of exporting it, to group things up a bit more

Rugvip reviewed (commented):
Thank you, good stuff indeed! 👍

kuangp commented:
@Rugvip updated per your review comments, thanks!

timbonicus reviewed (approved):
Thanks for the hard work on this, added tests, etc. It's really appreciated! 🙌

Rugvip reviewed (approved):
Thank you, let's :shipit:!
